pointcloud/pointcloud_file.go

```diff
@@ -545,69 +581,117 @@ func readPCDHelper(inRaw io.Reader, pctype PCType) (PointCloud, error) {
+}
+
+func readPCDASCII(in *bufio.Reader, header pcdHeader, pc PointCloud) (PointCloud, error) {
+	for i := 0; i < int(header.points); i++ {
+		pd, err := extractPCDPointASCII(in, header, i)
 		if err != nil {
 			return nil, err
 		}
-		err = pc.Set(pcPoint, data)
+		err = pc.Set(pd.P, pd.D)
 		if err != nil {
 			return nil, err
 		}
 	}
 	return pc, nil
 }
 
-func readPCDBinary(in *bufio.Reader, header pcdHeader, pc PointCloud) (PointCloud, error) {
-	var err error
+func getPCDMetaDataASCII(in bufio.Reader, header pcdHeader) (MetaData, error) {
+	meta := NewMetaData()
 	for i := 0; i < int(header.points); i++ {
-		pointBuf := make([]float64, 3)
-		colorData := NewBasicData()
-		for j := 0; j < 3; j++ {
-			buf, err := readBuffer(in, header, j)
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			pointBuf[j] = readFloat(binary.LittleEndian.Uint32(buf))
+		pd, err := extractPCDPointASCII(&in, header, i)
+		if err != nil {
+			return MetaData{}, err
 		}
-		point := r3.Vector{X: 1000. * pointBuf[0], Y: 1000. * pointBuf[1], Z: 1000. * pointBuf[2]}
+		meta.Merge(pd.P, pd.D)
+	}
+	return meta, nil
+}
 
-		if header.fields == pcdPointColor && !errors.Is(err, io.EOF) {
-			buf, err := readBuffer(in, header, 3)
-			if errors.Is(err, io.EOF) {
-				break
-			}
-			colorBuf := int(binary.LittleEndian.Uint32(buf))
-			colorData = NewColoredData(_pcdIntToColor(colorBuf))
+func extractPCDPointBinary(in *bufio.Reader, header pcdHeader) (PointAndData, error) {
+	var err error
+	pointBuf := make([]float64, 3)
+	colorData := NewBasicData()
+	for j := 0; j < 3; j++ {
+		buf, err := readBuffer(in, header, j)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return PointAndData{}, err
+		}
+		pointBuf[j] = readFloat(binary.LittleEndian.Uint32(buf))
+	}
+
+	// Converts PCD units (meters) to millimeters for RDK
+	point := r3.Vector{X: 1000. * pointBuf[0], Y: 1000. * pointBuf[1], Z: 1000. * pointBuf[2]}
+
+	if header.fields == pcdPointColor && !errors.Is(err, io.EOF) {
+		buf, err := readBuffer(in, header, 3)
+		if err != nil {
+			return PointAndData{}, err
+		}
+		colorBuf := int(binary.LittleEndian.Uint32(buf))
+		colorData = NewColoredData(_pcdIntToColor(colorBuf))
+	}
+
+	return PointAndData{P: point, D: colorData}, nil
+}
+
+func readPCDBinary(in *bufio.Reader, header pcdHeader, pc PointCloud) (PointCloud, error) {
+	for i := 0; i < int(header.points); i++ {
+		pd, err := extractPCDPointBinary(in, header)
+		if errors.Is(err, io.EOF) {
+			break
 		}
 		if err != nil {
 			return nil, err
 		}
-		err = pc.Set(point, colorData)
+		err = pc.Set(pd.P, pd.D)
 		if err != nil {
 			return nil, err
 		}
 	}
 	return pc, nil
 }
 
+func getPCDMetaDataBinary(in bufio.Reader, header pcdHeader) (MetaData, error) {
+	meta := NewMetaData()
+	for i := 0; i < int(header.points); i++ {
+		pd, err := extractPCDPointBinary(&in, header)
+		if err != nil {
+			return MetaData{}, err
+		}
+
+		meta.Merge(pd.P, pd.D)
+	}
+	return meta, nil
+}
```

Comment:
yes, I think combining the getPCDMetaData functions into one will make all right sense